### DISCUSSION TITLE
yup field.resolve is not a function

### DISCUSSION THREAD

**Discussion by andrconstruction** [react-hook-form/react-hook-form#2144] [Q&A] [answered]
Hello, please help.  I use the react hooks form with the yup validator. If no default values all works as expected, but if I set async default data for form, then after onsubmit I see this  error "yup field.resolve is not a function " witing an error object

code for a resolver set  :

   const schema = yup.object().shape( {
        isExistUser : (userId !== null),
        password : yup
            .string()
            .when( "isExistUser", {
                is : true,
                otherwise : yup.string().required().min( 8 )
            } ),
        username : yup.string().min( 2 ).max( 125 ).required(),
        display_name : yup.string().min( 2 ).max( 125 ).required(),
        email : yup.string().min( 2 ).max( 125 ).required(),
        work_phone : yup.string().min( 0 ).max( 4 ).required(),
        mobile_phone : yup.string().min( 1 ).max( 4 ),
        gender : yup.bool(),

    } );

    const methods = useForm( { user,  resolver : yupResolver(schema)} )

What should I dig?

**Comment by bluebill1049** [comment, marked as answer]
can you provide a codesandbox with this issue?

**Reply by andrconstruction** [reply]
https://codesandbox.io/s/0ykoc?file=/src/forms/UserContactsForm.js

**Reply by andrconstruction** [reply]
> can you provide a codesandbox with this issue?

blubill can you review sandbox code?

**Reply by bluebill1049** [reply]
what's the steps to reproduce the problem? can you reduce the scope of the code as well as save my time to look through the root cause?

**Reply by andrconstruction** [reply]
> what's the steps to reproduce the problem? can you reduce the scope of the code as well as save my time to look through the root cause?

Click the button "getExistsUser" and then click submit, within the console will be  filled a dump of  methods and if open last of them , an attribute errors would be contains an error

**Reply by bluebill1049** [reply]
<img width="652" alt="Screen Shot 2020-07-14 at 3 25 40 pm" src="https://user-images.githubusercontent.com/10513364/87387545-47719f00-c5e6-11ea-948c-ef1285e2e8b5.png">
i don't see the error you referred above.

**Reply by andrconstruction** [reply]
If an user exists and the form filled by a data

![image](https://user-images.githubusercontent.com/7435330/87390792-373fcc80-c5b2-11ea-9c98-6829e0c588a1.png)

**Reply by andrconstruction** [reply]
Otherwise when the form is empty, the normal behaviour
![image](https://user-images.githubusercontent.com/7435330/87390897-6b1af200-c5b2-11ea-91c3-7aa9a1346849.png)

**Reply by bluebill1049** [reply]
couple questions:
1. how did you reset the form 
2. are you using `reset` API, 
3. what's the value that you are trying to reset? any value is not part of the form.

**Reply by andrconstruction** [reply]
that as I use reset :

  useEffect(() => {
    reset(user);
    console.log("userid3");
    console.log(user.id);
  }, [user, reset]);

user object :

 const fakeUser = {
    id: 12,
    username: "vasya",
    display_name: "Vasiliy Vasiliy",
    removed: 0,
    email: "vasys@vasya",
    birth_day: "20.02.1990",
    gender: true,
    job: "",
    work_phone: "1445",
    mobile_phone: "",
    picture: "",
    state: 1,
    version: 1
  };

**Reply by bluebill1049** [reply]
is ther any `input` is not part of the registered field?

**Reply by andrconstruction** [reply]
> is ther any `input` is not part of the registered field?

No, all fields use Controller or ref={register}

**Reply by andrconstruction** [reply]
I simplify form till 5 inputs, and no luck. Why validation works with empty form but not works with filled? I suspect somehow happening within reset a form

**Reply by bluebill1049** [reply]
that's what i suspect as well, can you try to strip some of the data and see if that change anything.

**Reply by andrconstruction** [reply]
> that's what i suspect as well, can you try to strip some of the data and see if that change anything.

cutted but not helped, remaining only two fields in the user object : username and display_name, with same effect

**Reply by andrconstruction** [reply]
should I rebind yup Validation after reset?

**Reply by bluebill1049** [reply]
don't think so. this is strange.

**Reply by bluebill1049** [reply]
there are so many `reset(user);` in side your app as well, like I said it's' easier to reproduce in a simple codesandbox instead of dump the whole form/project.

**Reply by andrconstruction** [reply]
Leaves only one inside UserBaseForm, second part of form I have removed

**Reply by andrconstruction** [reply]
Can this problem related with nested components?

**Reply by bluebill1049** [reply]
I dont understand why you have so many reset()

**Reply by andrconstruction** [reply]
> I dont understand why you have so many reset()

At the moment I have only one reset, how should I use it within FormProvider if I use nested components?

**Reply by andrconstruction** [reply]
This last edited codesandbox
https://codesandbox.io/s/staging-architecture-oz72g?file=/src/service/UserContext.js

**Reply by andrconstruction** [reply]
It containes only one reset

**Reply by andrconstruction** [reply]
> > I dont understand why you have so many reset()
> 
> At the moment I have only one reset, how should I use it within FormProvider if I use nested components?

Is RHF connect the reset func from one nested components to other? Where is it must placed correctly?

**Reply by bluebill1049** [reply]
u should only need one reset to reset the form.

**Reply by andrconstruction** [reply]
> u should only need one reset to reset the form.

It I understand already, but where within FormProvider, have you example? As I told at this code contains only one reset :
https://codesandbox.io/s/staging-architecture-oz72g?file=/src/service/UserContext.js

**Reply by bluebill1049** [reply]
do you have any more reset at your child component's useEffect?

**Reply by andrconstruction** [reply]
No more resets. At this moment only one

**Reply by andrconstruction** [reply]
But the error the same

**Reply by bluebill1049** [reply]
not sure if this issue is related to your problem: https://github.com/react-hook-form/react-hook-form/issues/2222

**Reply by bluebill1049** [reply]
i am investigating the issue now.

**Reply by bluebill1049** [reply]
never mind it's not.

**Reply by andrconstruction** [reply]
It is a bug or I did something wrong?

**Reply by bluebill1049** [reply]
it's not a bug. I will try to reproduce a simple codesandbox fo you anyhow.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/inspiring-clarke-dfypp?file=/src/App.js
here is a working version with `reset`

i do see some potential problem with your schema it has variable and its inline as well.

**Reply by andrconstruction** [reply]
Yes indeed, problem with a conditional field. Thank you very much. Gorgeous support. Will connect with yup supports.
Also I think will be better add some documentation or code examples for complex form with nested components and async data.
Thanks!
